### ISSUE TITLE
Tests

### ISSUE THREAD

**Issue by rbumm** [lassoan/SlicerTotalSegmentator#1]
This is a nice addon!

**Desktop Windows RTX 3070 Ti**

Test with Slicer 5.3 preview:

\-> Working well, would automatic threedview generation and centering be helpful?  

Test with Slicer 5.0.3:

 -> TotalSegmentator runs from extension but does not create segments

**Laptop Windows RTX 1060**

Test with a fresh Slicer 5.3 preview:

error message (details follow) 

Test with Slicer 5.0.3:

 error message (details follow)

**Comment by lassoan** [comment]
Thanks a lot for testing!

Slicer 5.0.3: It will be discontinued (no extension updates will be provided by the factory machines) in a few days when Slicer-5.2 is released. So, the extension can only be supported for Slicer-5.2 or above.

Laptop with RTX 1060: I guess PyTorch does not find CUDA, and/or it runs out of CPU or GPU memory. The error message may help.

**Comment by lassoan** [comment]
FYI, I've tested it on a Surface laptop 3 (without a GPU) with 16GB RAM and computation (in fast mode) was successfully completed.
- CTACardio data set, in 8 minutes (this was the very first run, time included some one-time downloads)
- Panoramix-cropped data set 1 minute.
- CTACardio data set in 4.5 minutes

**Comment by lassoan** [comment]
> would automatic threedview generation and centering be helpful?  

I've added a "Show 3D" button

**Comment by rbumm** [comment]
> I've added a "Show 3D" button

Works well, but lets the output segmentation field just grow big. Automatic 3D centering would be nice.    

![](https://user-images.githubusercontent.com/18140094/204044752-c7f78ced-20dd-4dcc-9ce2-973c5f179de2.png)

**Comment by lassoan** [comment]
Thanks for testing. I've updated the layout, please check now.

**Comment by rbumm** [comment]
Now all good. 

![](https://user-images.githubusercontent.com/18140094/204050129-cb466ff7-39d9-4a3f-ad9f-5aa517c89f02.png)

**Comment by rbumm** [comment]
I tested the **Laptop Windows RTX 1060** again with a fresh 5.3 install - it works fine in --fast mode without errors.

It would be nice to have the option to update the TotalSegmentator install from GitHub if the user wants to use the latest version - or is that distracting / too complicated to understand for users?

**Comment by lassoan** [comment]
Thanks for testing, it is great to hear that it all works.

We could add a button in an "Advanced" section that displays information about the currently installed version and allows update. I add an issue to keep track of this idea.

**Comment by lassoan** [comment]
I've added buttons to the Advanced section to check current version and upgrade to latest version.

**Comment by rbumm** [comment]
> I've added buttons to the Advanced section to check current version and upgrade to latest version.

Works great, maybe we could catch the exception when the user presses "Get version information" and has not pressed "Apply" ever - when Totalsegmentator is not installed yet.

**Comment by rbumm** [comment]
Laptop Test Windows RTX 1060, fresh 5.3 install incl PyTorch:  

First "Apply" runs into 

```
ERROR: Failed to compute results.

Command '['C:/Users/Rudolf/AppData/Local/NA-MIC/Slicer 5.3.0-2022-11-26/bin/../bin\\PythonSlicer.EXE', '-m', 'pip', 'install', 'git+https://github.com/wasserth/TotalSegmentator.git']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "C:/Users/Rudolf/Documents/MySlicerExtensions/SlicerTotalSegmentator/TotalSegmentator/TotalSegmentator.py", line 236, in onApplyButton
    self.logic.setupPythonRequirements()
  File "C:/Users/Rudolf/Documents/MySlicerExtensions/SlicerTotalSegmentator/TotalSegmentator/TotalSegmentator.py", line 316, in setupPythonRequirements
    slicer.util.pip_install("git+https://github.com/wasserth/TotalSegmentator.git")
  File "C:\Users\Rudolf\AppData\Local\NA-MIC\Slicer 5.3.0-2022-11-26\bin\Python\slicer\util.py", line 3574, in pip_install
    _executePythonModule('pip', args)
  File "C:\Users\Rudolf\AppData\Local\NA-MIC\Slicer 5.3.0-2022-11-26\bin\Python\slicer\util.py", line 3536, in _executePythonModule
    logProcessOutput(proc)
  File "C:\Users\Rudolf\AppData\Local\NA-MIC\Slicer 5.3.0-2022-11-26\bin\Python\slicer\util.py", line 3505, in logProcessOutput
    raise CalledProcessError(retcode, proc.args, output=proc.stdout, stderr=proc.stderr)
subprocess.CalledProcessError: Command '['C:/Users/Rudolf/AppData/Local/NA-MIC/Slicer 5.3.0-2022-11-26/bin/../bin\\PythonSlicer.EXE', '-m', 'pip', 'install', 'git+https://github.com/wasserth/TotalSegmentator.git']' returned non-zero exit status 1.
```

The second "Apply" succeeds.

**Comment by lassoan** [comment]
@rbumm I've implemented a workaround for the SimpleITK issue (modify the SimpleITK requirements in TotalSegmentator package METADATA file). With this, installation seems to work robustly. It would be nice if you could give it a try. Use a completely fresh install of Slicer 5.2.1 (or Slicer 5.3.0 downloaded 2022-11-30 or later).

**Comment by rbumm** [comment]
@lassoan I did 

a fresh install of 5.2.1 on the RTX 1060 laptop
could not find TotalSegmentator in the extensions manager yet so installed it in modules after pulling your latest changes
installed Lung CT Analyzer in modules but did not run it
installed PyTorch from the extensions manager
installed ExtraEffects from the extensions manager
installed Wrapsolidify from the extensions manager
Did not run Lung CT Segmenter or Analyzer
 
loaded Chest CT Sample
switched to TotalSegmentator
checked "Fast" and pressed apply
the module prompted me to install PyTorch please wait
PyTorch install ok, 

![image](https://user-images.githubusercontent.com/18140094/204731964-a38b57f8-0209-43cc-961a-2edff76d0f47.png)

Second or third "Apply": expected result

**Comment by rbumm** [comment]
Then I do a "lungmask" Segmentation with Lung CT Segmenter and SimpleITK: Good result

![image](https://user-images.githubusercontent.com/18140094/204733410-e9e654e5-02f7-4100-b04a-a48ca224a163.png)

**Comment by rbumm** [comment]
Next follows an update of TotalSegmentator

![image](https://user-images.githubusercontent.com/18140094/204733985-90955ef2-a351-4788-a4b5-012ad71bd992.png)

(I can not see the update progress & messages in the textbux, because the textbox has no scrollbars)

**Comment by rbumm** [comment]
Then I do another "lungmask" segmentation with SimpleITK, which works nice. 

![image](https://user-images.githubusercontent.com/18140094/204734561-1a9bc8de-d9e9-4a60-b20a-476f101b0de8.png)

So I think the aforementioned SimpleITK problem is solved which is great, but we still have a glitch remaining upon first TotalSegmentator run

**Comment by rbumm** [comment]
After an unsuccessful morning test, I tried this again and reduced any preinstallations. 

Just did: 

*   Install a fresh 5.2.1 
*   Install PyTorch extension
*   Include SlicerTotalSegmentator in the module path
*   Loaded CTChest
*   Checked "Fast"
*   Pressed "Apply"

All worked well, after some minutes I got

![](https://user-images.githubusercontent.com/18140094/204790623-25c1cc1e-f35b-4c14-8502-d88bab0b5a91.png)

without an error message being raised. 

Output: 

```
PyTorch Python package is required. Installing... (it may take several minutes)
Matplotlib Python package is required. Installing...
TotalSegmentator Python package is required. Installing... (it may take several minutes)
TotalSegmentator installation is completed successfully.
Processing started
Writing input file to C:/Users/Rudolf/AppData/Local/Temp/Slicer/__SlicerTemp__2022-11-30_12+53+13.236/total-segmentator-input.nii
Creating segmentations with TotalSegmentator AI...
Total Segmentator arguments: ['-i', 'C:/Users/Rudolf/AppData/Local/Temp/Slicer/__SlicerTemp__2022-11-30_12+53+13.236/total-segmentator-input.nii', '-o', 'C:/Users/Rudolf/AppData/Local/Temp/Slicer/__SlicerTemp__2022-11-30_12+53+13.236/segmentation', '--ml', '--task', 'total', '--fast']
WARNING: In D:\a\1\sitk-build\ITK\Modules\Core\Common\src\itkObjectFactoryBase.cxx, line 597
Possible incompatible factory load:
Running itk version :
itk version 5.1.2
Loaded factory version:
itk version 5.3.0
Loading factory:
C:/Users/Rudolf/AppData/Local/NA-MIC/Slicer 5.2.1/bin/../lib/Slicer-5.2/ITKFactories\MRMLIDIOPlugin.dll


WARNING: In D:\a\1\sitk-build\ITK\Modules\Core\Common\src\itkObjectFactoryBase.cxx, line 597
Possible incompatible factory load:
Running itk version :
itk version 5.1.2
Loaded factory version:
itk version 5.3.0
Loading factory:
C:/Users/Rudolf/AppData/Local/NA-MIC/Slicer 5.2.1/bin/../lib/Slicer-5.2/ITKFactories\MRMLIDIOPlugin.dll


Error ImageIO factory did not return an ImageIOBase: MRMLIDImageIO


Please cite the following paper when using nnUNet:

Isensee, F., Jaeger, P.F., Kohl, S.A.A. et al. "nnU-Net: a self-configuring method for deep learning-based biomedical image segmentation." Nat Methods (2020). https://doi.org/10.1038/s41592-020-01008-z


If you have questions or suggestions, feel free to open an issue at https://github.com/MIC-DKFZ/nnUNet

preprocessing C:\Users\Rudolf\AppData\Local\Temp\nnunet_tmp_tok03dft\s01.nii.gz
using preprocessor GenericPreprocessor
before crop: (1, 116, 130, 130) after crop: (1, 116, 130, 130) spacing: [3. 3. 3.]

no resampling necessary
no resampling necessary
before: {'spacing': array([3., 3., 3.]), 'spacing_transposed': array([3., 3., 3.]), 'data.shape (data is transposed)': (1, 116, 130, 130)}
after:  {'spacing': array([3., 3., 3.]), 'data.shape (data is resampled)': (1, 116, 130, 130)}

(1, 116, 130, 130)
This worker has ended successfully, no errors to report
WARNING: In D:\a\1\sitk-build\ITK\Modules\Core\Common\src\itkObjectFactoryBase.cxx, line 597
Possible incompatible factory load:
Running itk version :
itk version 5.1.2
Loaded factory version:
itk version 5.3.0
Loading factory:
C:/Users/Rudolf/AppData/Local/NA-MIC/Slicer 5.2.1/bin/../lib/Slicer-5.2/ITKFactories\MRMLIDIOPlugin.dll


Error ImageIO factory did not return an ImageIOBase: MRMLIDImageIO


Please cite the following paper when using nnUNet:

Isensee, F., Jaeger, P.F., Kohl, S.A.A. et al. "nnU-Net: a self-configuring method for deep learning-based biomedical image segmentation." Nat Methods (2020). https://doi.org/10.1038/s41592-020-01008-z


If you have questions or suggestions, feel free to open an issue at https://github.com/MIC-DKFZ/nnUNet

force_separate_z: None interpolation order: 0
Using 'fast' option: resampling to lower resolution (3mm)
Resampling...
  Resampled in 2.43s
Predicting...
  Predicted in 16.30s
Resampling...
Saving segmentations...
  Saved in 0.65s
Importing segmentation results...
Cleaning up temporary folder...
Processing completed in 34.78 seconds

Processing finished.
```

**Comment by rbumm** [comment]
Then, I installed the necessary extensions for Lung CT Segmenter, included it in the module path, and ran it, unfortunately, I got the SimpleITK error again.

**Comment by lassoan** [comment]
Thanks for testing, I was able to reproduce the problem. Let's continue the discussion in a dedicated issue -> #4
